### DISCUSSION TITLE
Store scoped to component and its descendants

### DISCUSSION THREAD

**Discussion by donmccurdy** [pmndrs/zustand#1214] [Q&A]
Defining a component, `Forest.jsx`, composed of many descendant components, and Forest.jsx defines a store used by those descendants:

```javascript
import create from 'zustand'

const useWeatherStore = create((set) => ({
  rainfall: 0,
  increase: () => set((state) => ({ rainfall: state.rainfall + 1 })),
  decrease: () => set((state) => ({ rainfall: state.rainfall - 1 })),
}))

function Forest() {
  const rainfall = useWeatherStore((state) => state.rainfall)
  return (
    <Fragment>
      <Tree/>
      <Tree/>
      <Deer/>
      <Moose/>
      <Bear/>
    </Fragment>
  );
}
```

At runtime, outside the scope of my React codebase, the Forest component is imperatively instantiated multiple times in the DOM:

```js
ReactDOM.hydrate(
  <React.StrictMode>
    <Forest />
  </React.StrictMode>,
  document.querySelector('#forest-1')
);

ReactDOM.hydrate(
  <React.StrictMode>
    <Forest />
  </React.StrictMode>,
  document.querySelector('#forest-2')
);
```

In this situation the two instances of my Forest component share a store, and (for that matter) anything else declared in global scope, like the `ZustandContext` in [this example](https://github.com/pmndrs/zustand/discussions/566#discussioncomment-1383409). If I want each forest to have an independent "rainfall" store, this is now broken.

Is it possible to ensure each store is scoped only to this instance of the Forest component, and not to the root scope of any particular module? I read about [createContext](https://github.com/pmndrs/zustand/blob/main/docs/zustand-v3-create-context.md) and this sounds promising, but it also sounds like it will be removed in v5, is there a better way to handle this?

Thank you!

**Comment by dai-shi** [comment]
Please check #1180 for the migration guide from v3 createContext to v4 createStore.
